source/module_base/ndarray.h

```diff
@@ -0,0 +1,284 @@
+class NDArray
+{
+    // align with STL container implementation, there are several functions compulsory to be implemented
+    // constructor: default, copy, move, initializer_list
+    // operator: =, ==, !=, <, <=, >, >=
+    // iterator: begin, cbegin, end, cend
+    // capacity: size, empty, max_size, reserve, shrink_to_fit
+    // element access: [], at, front, back, data
+    // modifiers: clear, insert, erase, push_back, pop_back, resize, swap
+    // allocator: get_allocator
+public:
+    // constructors
+    /**
+     * @brief Construct a new NDArray object
+     * 
+     */
+    NDArray() : data_(), shape_() {}
+    // initializer_list constructor, size_t
+    NDArray(std::initializer_list<size_t> il) : shape_(il), data_(std::accumulate(shape_.begin(), shape_.end(), 1, std::multiplies<size_t>())) {}
+    // initializer_list constructor, int, cast to size_t
+    NDArray(std::initializer_list<int> il) : shape_(il.begin(), il.end()), data_(std::accumulate(shape_.begin(), shape_.end(), 1, std::multiplies<size_t>())) {}
+    // variadic template constructor, (delegate constructor)
+    template<typename... Args>
+    NDArray(Args... args) : NDArray({args...}) {}
+    // copy constructor
+    NDArray(const NDArray& other) : data_(other.data_), shape_(other.shape_) {}
+    // move constructor
+    NDArray(NDArray&& other) : data_(std::move(other.data_)), shape_(std::move(other.shape_)) {}
+
+    // destructor
+    ~NDArray() {}
+
+    // operators
+    /**
+     * @brief = operator, copy assignment
+     * 
+     * @param other 
+     * @return NDArray& 
+     */
+    NDArray& operator=(const NDArray& other)
+    {
+        if (this != &other)
+        {
+            data_ = other.data_;
+            shape_ = other.shape_;
+        }
+        return *this;
+    }
+    /**
+     * @brief = operator, move assignment
+    */
+    NDArray& operator=(NDArray&& other)
+    {
+        if (this != &other)
+        {
+            data_ = std::move(other.data_);
+            shape_ = std::move(other.shape_);
+        }
+        return *this;
+    }
+
+    /**
+     * @brief == operator
+     * 
+     * @param other 
+     * @return true if the data and shape are the same
+     * @return false otherwise
+     */
+    bool operator==(const NDArray& other) const
+    {
+        return data_ == other.data_ && shape_ == other.shape_;
+    }
+    /**
+     * @brief != operator
+     * 
+     * @param other 
+     * @return true if the data and shape are different
+     * @return false otherwise
+     */
+    bool operator!=(const NDArray& other) const
+    {
+        return !(*this == other);
+    }
+    // other operators are not generally supported
+
+    // element access
+    /**
+     * @brief at function
+     * 
+     * @tparam Args 
+     * @param args indices of the element
+     * @return T& 
+     */
+    template<typename... Args>
+    T& at(Args... args)
+    {
+        size_t idx = index(args...);
+        return data_.at(idx);
```

Comment:
looks like this could be a one-liner:
```suggestion
        return data_.at(index(args...));
```
or, if bound is already checked in index(), might be just ```data_[index(args...)]```